tools/rustfmt/srcs/main.rs

```diff
@@ -0,0 +1,207 @@
+}
+
+/// Build a list of Bazel targets using the `rustfmt_aspect` to produce the
+/// arguments to use when formatting the sources of those targets.
+fn build_rustfmt_targets(options: &Config, targets: &Vec<String>) {
+    let build_args = vec![
+        "build",
+        "--aspects=@rules_rust//rust:defs.bzl%rustfmt_aspect",
+        "--output_groups=+rustfmt",
+    ];
+
+    let child = Command::new(&options.bazel)
+        .current_dir(&options.workspace)
+        .args(build_args)
+        .args(targets)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::inherit())
+        .spawn()
+        .expect("Failed to spawn command");
+
+    let output = child
+        .wait_with_output()
+        .expect("Failed to wait on spawned command");
+
+    if !output.status.success() {
+        std::process::exit(output.status.code().unwrap_or(1));
+    }
+}
+
+/// Run rustfmt on a set of Bazel targets
+fn apply_rustfmt(options: &Config, targets: &Vec<String>) {
+    // Ensure the targets are first built and a manifest containing `rustfmt`
+    // arguments are generated before formatting source files.
+    build_rustfmt_targets(&options, &targets);
+
+    for target in targets.iter() {
+        // Replace any `:` characters and strip leading slashes
+        let target_path = target.replace(":", "/").trim_start_matches("/").to_owned();
+
+        // Load the manifest containing rustfmt arguments
+        let rustfmt_config = parse_rustfmt_manifest(
+            &options
+                .workspace
+                .join("bazel-bin")
+                .join(format!("{}.rustfmt", &target_path)),
+        );
+
+        // Ignore any targets which do not have source files. This can
+        // occur in cases where all source files are generated.
+        if rustfmt_config.sources.is_empty() {
+            continue;
+        }
+
+        // Run rustfmt
+        let status = Command::new(&options.rustfmt)
+            .current_dir(&options.workspace)
+            .arg("--edition")
+            .arg(rustfmt_config.edition)
+            .args(rustfmt_config.sources)
+            .status()
+            .expect("Failed to run rustfmt");
+
+        if !status.success() {
+            std::process::exit(status.code().unwrap_or(1));
+        }
+    }
+}
+
+/// A struct containing details used for executing rustfmt.
+#[derive(Debug)]
+struct Config {
+    /// The path of the Bazel workspace root.
+    pub workspace: PathBuf,
+
+    /// The Bazel executable to use for builds and queries.
+    pub bazel: PathBuf,
+
+    /// The rustfmt binary from the currently active toolchain
+    pub rustfmt: PathBuf,
+
+    /// The rustfmt config file containing rustfmt settings.
+    /// https://rust-lang.github.io/rustfmt/
+    pub config: PathBuf,
+
+    /// An optional command line flag used to control what targets
+    /// to format. Users are expected to either pass a Bazel label
+    /// or a package pattern (`//my/package/...`).
+    pub package: Option<String>,
```

Comment:
It would be nice if this were a `Vec<String>` rather then `Option<String>` so that if you have multiple things to fmt, you can do so in one invocation.

I think it would just be a matter of changing the type here, changing the `nth(1)` below to a `skip(1).collect()`, and a small tweak to `query_rustfmt_targets`?

Reply:
Ok, updated, you can now pass a list of targets or query scopes.